pkg/ingester/metadata_map.go

```diff
@@ -0,0 +1,153 @@
+package ingester
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/model"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+)
+
+type metadataMap struct {
```

Comment:
This is as simple as it can be. There's no sharding and no concurrent iteration - the volume of metadata is very low and IMHO does not justify the increased complexity. Please let me know if you differ and I'll be happy to refactor to a more performant version.